Support for external/HD textures

rafccq opened the pull request:
### Summary of changes:
- Removed alignment from the original textures data, which caused some extra junk (padding bytes) to be uploaded
- Removed the UV adjustments in the fast3d shader that deals with the junk data
- Added new GDL command to identify textures in the renderer (tex number/font/embedded in model)
- Added new GDL TLUT command to set a pointer directly to palette colors
- Preprocessing now inserts the texture-identification command for embedded model textures
- Renderer updated to load and cache external textures when available and enabled
- On init the texture manager will iterate the external tex files and build lookup arrays

### How to use:
External textures live in a new folder: `ext_tex` inside the game’s data folder.

#### General textures
- Place files directly in `ext_tex`, named with a 4-digit lowercase hex (e.g. `005a.png`, `0c18.jpg`)
~Note that the extension should be exactly 3 characters, so `jpg` instead of `jpeg`~
- Supported formats: PNG, JPG, TGA, BMP, PSD, GIF, HDR, PIC, PNM (via stb_image)
- Currently only PNG is tested

#### Font textures
- There are 5 fonts in the game: `fonthandelgothicsm` `fonthandelgothicmd` `fonthandelgothicxs` `fonthandelgothiclg` and  `fontnumeric`
- Each character goes in `ext_tex/<fontname>/` named as `(ascii_code - 0x21).png`
  - Example: 'A' (0x41) → 20.png
- Outlines go in `ext_tex/<fontname>/outlines/` with the same filenames
- Japanese fonts are not yet supported but will be soon

Here's a table for reference:
```
00 !    0f 0    1e ?    2d N    3c ]    4b l
01 "    10 1    1f @    2e O    3d ^    4c m
02 #    11 2    20 A    2f P    3e _    4d n
03 $    12 3    21 B    30 Q    3f `    4e o
04 %    13 4    22 C    31 R    40 a    4f p
05 &    14 5    23 D    32 S    41 b    50 q
06 '    15 6    24 E    33 T    42 c    51 r
07 (    16 7    25 F    34 U    43 d    52 s
08 )    17 8    26 G    35 V    44 e    53 t
09 *    18 9    27 H    36 W    45 f    54 u
0a +    19 :    28 I    37 X    46 g    55 v
0b ,    1a ;    29 J    38 Y    47 h    56 w
0c -    1b <    2a K    39 Z    48 i    57 x
0d .    1c =    2b L    3a [    49 j    58 y
0e /    1d >    2c M    3b \    4a k    59 z
```

#### Model embedded textures
- Place inside a subfolder named after the model
- Filenames should be the lower 4–5 hex digits of the texture ID

Example:
```
PinstfrontdoorZ
  0b40.png
  0088.png
  ...

Ca51guardZ
  005E0.png
  1AF40.png
  ...
```

Commit:
port-ext-tex: removed alignments from textures; renderer: always use tex size calculated from tiles

Commit:
port-ext-tex: custom LOADTLUT command

to avoid having to calculate the LUT offset into the texture data, pass it directly to the renderer.
The previous implementation is still needed as some code will issue it

Commit:
port-ext-tex: removed redundant function in tex.c

Commit:
port-ext-tex: UVs: map to range [0, 1], removed adjustments in shader

I'm not 100% sure about the `cms &= ~G_TX_CLAMP;` lines but it seems fine

Commit:
port-ext-tex: fixed texdecompress bug described in the comment

Commit:
port-ext-tex: remove alignment from model embedded textures

Commit:
port-ext-tex: rename `struct tex.unk0a-> numcolors`

Commit:
port-ext-tex: custom gbi cmd to identify textures on the renderer

Commit:
port-ext-tex: external textures management

Commit:
port-ext-tex: renderer: load external textures

Commit:
port-ext-tex: added gdl cmd to identify the texture in models with embedded textures

Commit:
port-ext-tex: call cmd to identify textures from various parts in the game code

Commit:
port-ext-tex: texture config gdls: identify textures/fixed tile size params

Commit:
port-ext-tex: enable/disable external textures in real time; video menu option

Commit:
Merge branch 'fgs-port' into port-ext-textures

Commit:
port-ext-tex: fixed `S_ISFMT’ undeclared` on linux

Commit:
port-ext-tex: fixed explosions

Shepard-N7 commented:
Hello, I would like to know if there is any way to extract the textures from the game's ROM to see the name of each one and, in that way, rename an HD texture pack and perform tests.

rafccq commented:
If you really need I can provide you a script to extract the textures. 
However there is currently an existing HD texture pack whose author is working on renaming them.
There are also 2 other ways of identifying/extracting textures:
- via the "Perfect Gold Editor" or PGE or
- via my Blender Addon (_pd_blendtools_ under my repos)

Commit:
port-ext-tex: tons of adjustments

- Got rid of the hardcoded model list, resolve the name/fileNum from the ROM
- Increased extensions size to include the null terminating char and support 4 letter extensions
- Look up models from a dynamically allocated (at init) list of `struct ModelTextures`
- Function to get the font name instead of keeping an array of structs
- Removed unused function `extLoadFontTex`

Commit:
port-ext-tex: renamed `ExternalTex` -> `ExtTexture`

Commit:
port-ext-tex: renderer: changed when to use tex_width/tex_width2 when calculating UVs

Commit:
port-ext-tex: use 5 digits for embedded model texture numbers

Commit:
port-ext-tex: use relative path to retrieve ext tex folder

Commit:
port-ext-tex: free texture data on cache reset

Commit:
port-ext-tex: changed texnum data size, properly allowing texture names to have more than 4 digits

Commit:
port-ext-tex: don't purge video texture cache from `bgunFreeGunMem`

Commit:
port-ext-tex: font outlines: introduced `id_mask` to let the renderer load a related texture

when used in 2-cycle mode, the associated texture will be loaded in the first cycle (using the masked id), and the regular texture in the 2nd cycle

Commit:
port-ext-tex: no more harcoded number of fonts

Commit:
port-ext-tex: bug fix in `texSelect` when tex is NULL

Commit:
port-ext-tex: properly handle the case where the external tex can't be loaded

- set the `req_comp` param in `stbi_load` to always convert the texture to RGBA24
- if the texture cant be loaded, upload a 1 pixel replacement for visual feedback

Commit:
port-ext-tex: fixed font outlines not being imported properly

rafccq commented:
Here it is. You will need python to use it.
To run it, simply extract the files to a folder, put your ROM inside it then run `python extract_tex.py`
There's a readme inside with a bit more detailed instructions.
[PD_extract_tex.zip](https://github.com/user-attachments/files/23272007/PD_extract_tex.zip)

parabolee commented:
I would love to help with the texture upgrades, anyway I can do that? Do we have the ability to rip the XBLA textures and use them as a starting point?

parabolee commented:
> We are currently working on renaming all textures from Xii Nyth's pack into this new format. If you want to help with that, join the N64 Vault Discord server. As for the XBLA textures we're not working on it for various reasons, one of them being the legality of it, it's probably not legal to distribute it.

Yeah I would love to help with that. How are you guys dividing the duties? Took a look myself to see if there was an easy way and it seems dividing all the images by size would be the best way to identify them against the images your tool pulls. I was already able to do some faces. I'll jump on there.

I did Joanna but the N64 HD textures had some overlap issues so I started working on my own version closer to the XBLA style rather than fix someone else's work (seemed rude). I already have a HD texture pack working by using Topaz Gigapixel to upscale all the ripped textures. But obviously the had done ones will be better in the long run.

Good point on the XBLA textures, I had not considered that. I am sure no one at MS or 4J really care about us using the textures after 15 years. BUT considering the work it will take to rip, sort and label them. Yeah it's probably not worth the time with the fact there is a real risk they may just have us take them down anyway.

rafccq commented:
oh I actually forgot to save it into the .ini
I'm working on new features related to HD textures such as mipmap generation/anisotropic filtering etc, I'll include it in there.

Commit:
Merge remote-tracking branch 'fgs/port' into port-ext-textures

Commit:
port-ext-tex: save gfx_external_textures_enabled in the config

Atari-Dude commented:
The textures from Xii-Nyth's pack have more or less been ready for months (UI and map textures look pretty much complete-- there's only been a few test HD character textures by folks like myself, though), so that's ready for when either this PR is merged to the main port here, or the functionality makes its way to the All In One mod. Whichever comes first. Still minor bugs with a couple textures (such as related to glass) with the HD texture pack on the PC port, but those will be ironed out once this external/HD texture support functionality has been merged here or AIO, I think.
